File: Python/023_MergeKSortedLists.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def mergeKLists(self, lists):
        """
        :type lists: List[ListNode]
        :rtype: ListNode
        """
        if len(lists) == 0:
            return None
            
        if len(lists) == 1:
            return lists[0]
            
        mid = len(lists) // 2
        left = self.mergeKLists(lists[:mid])
        right = self.mergeKLists(lists[mid:])
        #merge left and right
        dummy = ListNode(0)
        cur = dummy
        while left or right:
            if right == None or (left and left.val < right.val):
                cur.next = left
                left = left.next
            else:
                cur.next = right
                right = right.next
            cur = cur.next
            
        return dummy.next

File: Java/023_MergeKSortedLists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 * http://cuisuqiang.iteye.com/blog/2019157 priorityQueue用法
 */
public class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        if (lists.length == 0) {
            return null;
        }
        PriorityQueue<ListNode> q = new PriorityQueue<ListNode>(lists.length,
                    new Comparator<ListNode>(){
                        public int compare(ListNode a, ListNode b) {
                            if (a.val > b.val) {
                                return 1;
                            } else if (a.val == b.val) {
                                return 0;
                            } else {
                                return -1;
                            }
                        }
                    }); 
        for (ListNode list : lists) {
            if (list != null) {
                q.add(list);
            }
        }  
        
        ListNode head = new ListNode(0);
        ListNode p = head;
        
        while (q.size() > 0) {
            ListNode temp = q.poll();
            p.next = temp;
            if (temp.next != null) {
                q.add(temp.next);
            }
            p = p.next;
        }
        
        return head.next;
    }
}
